File: 204-count-primes/count-primes.java
class Solution {
    public int countPrimes(int n) {

        if(n == 0 || n == 1){return 0;}
        
        boolean[] comp = sieve(n);
        int cnt = 0;
        for(int i = 2; i < n; i++){
            if(!comp[i]){cnt++;}
        }

        return cnt;
    }

    public boolean[] sieve(int n) {
        boolean[] comp = new boolean[n]; // 0~n-1
        comp[0] = comp[1] = true;

        for(int i = 2; i * i < n; i++){
            if(!comp[i]){
                for(int j = i * i; j < n; j += i){comp[j] = true;}
            }
        }

        return comp;
    }
}

File: 204-count-primes/count-primes.kt
import java.util.ArrayList;

class Solution {
    fun countPrimes(n: Int): Int {
        
        if(n == 0 || n == 1){return 0}
        
        val compositeNumbers = BooleanArray(n) // 0~N
        compositeNumbers[0] = true
        compositeNumbers[1] = true

        var i = 2
        while(i * i < n){
            if(!compositeNumbers[i]){
                var j = i * i
                while(j < n){
                    compositeNumbers[j] = true
                    j += i
                }
            }
            i++
        }

        var cnt = 0

        for(idx in 0..<n){
            if(!compositeNumbers[idx]){cnt++}
        }

        return cnt
    }
}